ironfish/src/syncer.ts

```diff
@@ -123,16 +162,90 @@ export class Syncer {
-    // peer diversity so the highest work peer isn't overwhelmed
-    // as well as helping to make sure we don't get stuck with unstable peers
-    if (peers.length > 0) {
-      const peer = ArrayUtils.sampleOrThrow(peers)
+    if (peers.length === 0) {
+      return
+    }
+
+    // If there is only one valid peer to sync from, there is no point in
+    // measuring the connection so begin syncing immediately
+    if (peers.length === 1) {
+      this.startSync(peers[0])
+      return
+    }
+
+    const syncCandidates = ArrayUtils.shuffle(peers).slice(0, 8)
+
+    // If we have been syncing from a peer, we want to include this peer in the
+    // measurement. This will allow us to maintain a connection to a strong peer
+    // if we have found one.
+    if (
+      currentPeerIdentity &&
+      !syncCandidates.find((p) => p.getIdentity() === currentPeerIdentity)
+    ) {
+      const currentPeer = this.peerNetwork.peerManager.getPeer(currentPeerIdentity)
+      if (currentPeer && currentPeer.state.type === 'CONNECTED') {
+        syncCandidates.push(currentPeer)
+      }
+    }
+
+    const peerRtt = new Map<Identity, number>()
+
+    // Measure how long it takes to fetch the genesis block header from each
+    // peer as an estimate of connection quality
+    for (const peer of syncCandidates) {
+      if (this.state === 'stopped' || this.state === 'stopping') {
+        return
+      }
+
+      const peerIdentity = peer.getIdentity()
+      if (peerIdentity == null) {
+        continue
+      }
+
+      const start = BenchUtils.start()
+      try {
+        const response = await this.peerNetwork.getBlockHeaders(peer, 1, 1)
+        if (response.headers.length !== 1) {
+          this.logger.warn(`Peer ${peer.displayName} sent the wrong number of block headers`)
+          peer.punish(BAN_SCORE.MAX, 'invalid response')
+          continue
+        }
+        if (!response.headers[0].hash.equals(this.chain.genesis.hash)) {
+          this.logger.warn(`Peer ${peer.displayName} sent the wrong block header`)
+          peer.punish(BAN_SCORE.MAX, 'invalid response')
+          continue
+        }
+      } catch (e) {
+        this.logger.debug(
+          `Error while trying to measure peer '${
+            peer.displayName
+          }', skipping this peer: ${ErrorUtils.renderError(e)}`,
+        )
+        continue
+      }
+      const rtt = BenchUtils.end(start)
+      peerRtt.set(peerIdentity, rtt)
+    }
+
+    // Sort the peers by the round-trip-time of the block header request and get
+    // the fastest one to sync from
+    const fastestCandidateIdentity = [...peerRtt.entries()].sort((a, b) => a[1] - b[1])[0][0]
+    if (!fastestCandidateIdentity) {
+      return
+    }
+    const peer = this.peerNetwork.peerManager.getPeer(fastestCandidateIdentity)
+    if (peer) {
+      this.startSync(peer)
+    }
+  }
+
+  startSyncIfIdle(peer: Peer): void {
+    if (this.state === 'idle') {
```

Comment:
I think we'll want to make sure that `this.nextMeasureTime` is updated when starting to sync directly from this function, otherwise we'll end up with `nextMeasureTime` set to whatever it was last in the event loop. Though generally the event loop should always be running and keeping it up to date.